"Model Workflows" section in Guides (#209)
validmind/documentation: site/guide/set-up-model-workflows.qmd

---
title: "Set up model workflows"
date: last-modified
---

Workflows can be configured to match your organizational needs for overseeing model development, validation, or implementation activities.

Default workflows are provided for you as suggestions. You can customize the following default workflows:

- **Model documentation:** When documentation is created

> - **Inventory model:** When the model is created

Author: I grabbed these descriptions from the UI, so I'm just matching the wording there to minimize confusion.

<!-- 
- Model validation workflow is in missing as of May 2024
- These default workflows will eventually be replaced by a trigger you can define
 -->

 Setting up your workflow involves two major steps:

>  1. [Configuring workflow steps](#configure-workflow-steps) 

Author: Right, but if we match what the heading says, then the `:` doesn't make sense... (the introductory text leading up to it, I mean)

## Configure workflow steps

To add or edit workflow steps:

1. [Log in to the {{< var vm_platform >}}]({{< var vm_url >}}).

2. In the left sidebar, click [**{{< fa gear >}} Settings**]({{< var vm_url >}}/settings).

3. Under Workplace Settings, select **{{< fa arrow-right-arrow-left >}} Workflows**.

4. Under available workflows, select the process you'd like to add or edit the steps for:
 
    - To add a step, drag and drop a new step onto the canvas. Hover over the step until the **{{< fa ellipsis-vertical >}}** menu appears, then click on it and select **{{< fa gear >}} Configure**.
    - To edit a step, hover over the step until the **{{< fa ellipsis-vertical >}}** menu appears. Then click on it and select **{{< fa gear >}} Configure**.

5. After you're finished with step configuration, click **Save** to apply your changes before closing the configuration panel with **{{< fa x >}}**. 

    ![](configure-steps.gif){fig-alt="A gif workflow step configuration."}

    After you've configured a step, you can then [link your workflow together](#link-workflow-together).

::: {.callout-tip appearance="minimal"}

> Refer to **[Available step types](#available-step-types)** for information on available step types and details on how to set up conditional requirements.

Reviewer: GLOBAL: I would not bold links to subsections. This topic already has a ton of highlighting in it, so whatever extra visual clutter we can omit, we should.

:::

### Available step types

#### Status Change

- Transitions a status. 
- Requires [available resource statuses](customize-resource-statuses.qmd) associated with the workflow.

| Field | Description |
|---:|---|
| When these conditions are met (optional) | [Add conditional requirements](#add-conditional-requirements) to qualify for this step.  |
| Set status to (required) | Select the status to transition to. |

> : Status Change step configuration {.hover tbl-colwidths="[40,60]"}

Reviewer: ```suggestion
: **Status Change** step configuration {.hover tbl-colwidths="[40,60]"}
```

GLOBAL: References to UI elements you interact with should be bolded.

Do you know what the icon library is that we're using? We might be able to add the same to Quarto.

Reviewer: Also, does `{.hover tbl-colwidths="[40,60]"}` work for you? It doesn't seem to do anything when I tested locally.

Author: > Do you know what the icon library is that we're using? We might be able to add the same to Quarto.

I asked Jeffrey about the icons, they're not from Font Awesome (they're little images from Figma). I figured it was more trouble than it was worth to add images here.

> Also, does {.hover tbl-colwidths="[40,60]"} work for you? It doesn't seem to do anything when I tested locally.

It does, it adjusts the width effectively for me, not sure why it doesn't for you. [This is also how the official Quarto docs tell you to assign table widths.](https://quarto.org/docs/authoring/tables.html#column-widths)

Reviewer: > It does, it adjusts the width effectively for me, not sure why it doesn't for you. [This is also how the official Quarto docs tell you to assign table widths.](https://quarto.org/docs/authoring/tables.html#column-widths)

I figured it out — I'd never used `.hover` before but mainly because **I try to avoid tables wherever possible**.
Maybe my comments relates to our use of tables more generally — they can be problematic for screen readers and require special attention to be accessible ([How tables can make content less accessible](https://medium.com/@lizzie.g.bruce/accessibility-and-tables-c233f6b2b823)) 

In this case, I do think tables are helpful but they should only be used to represent data or parameters. Tables should not be used to format or group other content together. And even where a table might make sense, I would use a simple definition list before using a table, as it sidesteps accessibility issues altogether.

Author: That makes sense. Would you like me to change these to a list, or can we keep them as tables? I think the information here makes *sense* as a table, as they are key / value pairs, but I understand the argument against them.

Reviewer: I'd leave as is, but let's avoid a proliferation of tables as we expand our docs. "Use tables sparingly" would likely be a good addition to our style guide at some point.

#### User Action
- Creates a button that performs an action on behalf of the user. 
- Requires [available resource statuses](customize-resource-statuses.qmd) associated with the workflow.

| Field | Description |
|---:|---|
| Display action button when (optional) | [Add conditional requirements](#add-conditional-requirements) for displaying this step.  |
| Action button label (required) | Text that displays on your action button. |
| Request fields (optional) | Include [custom fields](manage-inventory-custom-fields.qmd) alongside your action if desired. Select the field, then click **+ Add Field** to insert.  |
| To set status (required) | Select the [status](customize-resource-statuses.qmd) to transition to. |
: User Action step configuration {.hover tbl-colwidths="[40,60]"}

::: {.grid}

::: {.g-col-7}

> ![](configure-user-action.png){fig-alt="A screenshot showing a User Action step configuration with a request field of Date On Hold."}

Reviewer: The screenshots are really helpful for this rather complex task, but they sometimes show extra stuff that adds visual complexity. E.g. here, I'd remove the extraneous stuff outside the modal/popup itself. 

Related, we should have a chat about borders around images. We don't have a standard for this and sometimes I've been adding them manually and sometimes with some CSS. In this PR, images don't have borders but they might benefit from them — ideally we sort this out globally via `styles.css` so they work for all images without having to edit them individually (other than removing manually added borders from existing images).

Author: I did already crop this screenshot, and I don't think making it smaller helps here — as a user, I'd expect to see the field in context with the other stuff so I could potentially locate it instead of looking for an element in isolation.

Reviewer: It's stuff like this that I am talking about around the edges which doesn't add context, it just looks like the screenshot needs to be edited. I would crop this and other screenshots to be just the modal itself, not the stray background lines or text behind it. 

<img width="536" alt="image" src="https://github.com/validmind/documentation/assets/15148011/02dc4a63-3e5f-4e03-bdeb-266aad32c0a7">

Author: Ohhh I see, sorry, I thought you meant the rest of the configuration box. I can definitely make the crop tighter.

:::

::: {.g-col-5}
![](user-action-custom-field.png){fig-alt="A screenshot showing the action transition panel with the Date On Hold custom field."}

> ::: {.callout-tip title="Example" appearance="minimal"}

Reviewer: FYI, Quarto also supports putting callouts into the right margin which might help clean up a visually very busy section in the docs:

::: {.column-margin}
EXISTING_CALLOUT
:::

Author: Interesting! I will try that out and see how it looks.

Author: That doesn't look as good for these sets of images, so I'm going with the old version.

A `Put On Hold` action that requests a value in the Date On Hold field, and transitions the resource status to On Hold.
:::

:::

:::

#### Approval

> - When your resource reaches the approval step, users in the selected Approval Group will receive your configured message. 

Reviewer: Is "Approval Group" a UI element? If so, it should be bolded. If not, it should be lowercased. (I also see this referred to as "Approval group" elsewhere, so some consistency sleuthing might be needed.)

> - Requires a User type [custom field](manage-inventory-custom-fields.qmd) to select from, and must be connected to both an approval and rejection **[Status Change](#status-change)** step.

Reviewer: ```suggestion
- Requires a **User** type [custom field](manage-inventory-custom-fields.qmd) to select from, and must be connected to both an approval and rejection **[Status Change](#status-change)** step.
```

I think that is the actual field name based on the link, so it should be bolded?

::: callout 

> Please note that users must be assigned to the Approval Group custom field before it is transitioned into the **Approval** workflow step in order for the message to be displayed to them. 

Reviewer: ```suggestion
Please note that you must assign users to the **Approval Group** custom field before it is transitioned into the **Approval** workflow step in order for the message to be displayed to them. 
```

Use active voice, George Orwell said so.

> Please note that users must be assigned to the Approval Group custom field before it is transitioned into the **Approval** workflow step in order for the message to be displayed to them. 

Reviewer: GLOBAL: Same here, there more unbolded UI elements here.

Author: Yeah, I'm still on the hill of "you can't click on these elements, so they shouldn't be bolded"... 😅

Reviewer: Let's chat later today and agree on what to do. My take is that since you are being asked to locate that section in the UI, you ARE interacting with it — highlighting this makes sense. We highlight other, non-UI texts for emphasis already.

Author: I made a quick change to the style guide as a suggestion, and edited the these workflow docs to reflect that!

::: {.callout-tip title="Example" appearance="minimal"}
A message shown to the Approval Group users requesting approval of model documentation.
:::

:::

:::

### Add conditional requirements

Conditional requirements can be configured for all three available step types:

| Step type | Conditional options |
|---:|---|
| Status Change | Under `When these conditions are met`, you are able to set both `AND` and `OR` conditions. |
| User Action | Under `Display action button when`, you are able to set both `AND` and `OR` conditions.  |
| Approval | Under `Threshold`, you are able to set the minimum percentage of approvers required for the resource to be approved. |
: Step type conditional options {.hover tbl-colwidths="[20,80]"}

For [Status Change](#status-change) and [User Action](#user-action) conditions, you're able to add a single independent **Rule** or a linked condition **Group**. These rules and groups can be nested if desired:

- Click **+ Rule** to add an independent rule.
- Click **+ Group** to add a linked group of rules that all must be true to qualify. 

## Link workflow together

To initiate the beginning of your workflow:

1. Locate the **Start** of your workflow.

> 2. Drag from the **.** at the bottom of **Start** to the top **{{< fa circle >}}** on the first step of your workflow.

Reviewer: ```suggestion
2. Drag from the period (**.**) at the bottom of **Start** to the top **{{< fa circle >}}** on the first step of your workflow.
```

This read like a broken sentence until I clued in. Not ideal, but I wonder if spelling it out isn't preferrable here?

### Link steps together

To link subsequent steps together:

1. Click **{{< fa circle >}}** above or below the earlier step and drag it to connect to the top **{{< fa circle >}}** on the subsequent step.

    You're also able to link several different steps together at various points in the workflow, including stages that may circle back to previous steps: 

    ![](linked-workflow.png){fig-alt="A screenshot showing an example of multiple steps linked together at various points in the default Inventory Model workflow."}

6. When you are finished, click **Save workflow** to apply your changes.

> #### Link Approval steps

Reviewer: ```suggestion
#### Link approval steps
```
